src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+                }
+                term error_tuple = term_alloc_tuple(2, &ctx->heap);
+                term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+                term_put_tuple_element(error_tuple, 1, posix_errno_to_term(errno, global));
+                return error_tuple;
+            } else {
+                return OK_ATOM;
+            }
+            // TODO add more as needed
+            // int flag = 1;
+            // int res = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int));
+            // if (UNLIKELY(res != 0)) {
+            //     AVM_LOGW(TAG, "Failed to set TCP_NODELAY.");
+            // }
+        } else {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+//
+// sockname
+//
+
+static term nif_socket_sockname(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_sockname\n");
+    UNUSED(argc);
+
+    VALIDATE_VALUE(argv[0], term_is_socket);
+
+    GlobalContext *global = ctx->global;
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), get_socket(argv[0]), socket_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct SocketFd *fd_obj = (struct SocketFd *) fd_obj_ptr;
+    if (fd_obj->fd) {
+        struct sockaddr_in addr;
+        socklen_t addrlen = sizeof(addr);
+        int res = getsockname(fd_obj->fd, (struct sockaddr *) &addr, &addrlen);
+
+        if (UNLIKELY(res != 0)) {
+            AVM_LOGE(TAG, "Unable to getsockname: fd=%i res=%i.", fd_obj->fd, res);
+            if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+                AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            }
+
+            term error_tuple = term_alloc_tuple(2, &ctx->heap);
+            term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+            term_put_tuple_element(error_tuple, 1, posix_errno_to_term(errno, global));
+            return error_tuple;
+        } else {
+            // {ok, #{addr => {a,b,c,d}, port => integer()}}
+            if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) + term_map_size_in_terms(2) + TUPLE_SIZE(4)) != MEMORY_GC_OK)) {
+                AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            } else {
+                term address = socket_tuple_from_addr(ctx, ntohl(addr.sin_addr.s_addr));
+                term port_number = term_from_int(ntohs(addr.sin_port));
+
+                term map = term_alloc_map(2, &ctx->heap);
+                term_set_map_assoc(map, 0, ADDR_ATOM, address);
+                term_set_map_assoc(map, 1, PORT_ATOM, port_number);
+                term return_value = port_create_tuple2(ctx, OK_ATOM, map);
+
+                return return_value;
+            }
+        }
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
```

Comment:
I understand we end up here if socket was closed. Is raising badarg the proper behavior?

Reply:
I changed these to ebadf``